design_docs/generalized_horizontal_interpolation.md

```diff
@@ -0,0 +1,99 @@
+
+<h2>Design solution: Higher-order interpolation <br>
+Date last modified: 2017/02/25 <br>
+Contributors: Xylar Asay-Davis</h2>
+
+The approach will be to create SCRIP files (or, in the future for greater flexibility perhaps ESMF grid/mesh files) for the source and destination grids, then to use `ESMF_RegridWeightGen` to generate a mapping file.  `ESMF_RegridWeightGen` supports 5 interpolation methods---bilinear, patch, nearestdtos, neareststod, and conserve---and we would likely support at least bilinear, neareststod and conserve, and perhaps all 5.  The destination grid will be specified either by reading values from `lat` and `lon` coordinates of a NetCDF file or through config file options `lat` and `lon` that are typically expressions involving `numpy.arange` or `numpy.linspace`.
+
+Then, `ncremap` will be used to remap the desired list of variables from an MPAS NetCDF file to the desired destination grid.
+
+<h2>Design solution: Interpolation should handle periodic boundaries <br>
+Date last modified: 2017/02/25 <br>
+Contributors: Xylar Asay-Davis</h2>
+
+For now, periodic boundaries (except for the obvious one at +/- 180 longitude) will not be supported.  It appears that ESMF grid files do include support for periodic boundaries so the current solution should be relatively easy to extend to periodic boundaries in the future.
+
+<h2>Design solution: Interpolation should handle Cartesian meshes <br>
+Date last modified:  2017/02/25 <br>
+Contributors: Xylar Asay-Davis</h2>
+
+ESMF unstructured mesh files seem to support Cartesian coordinates.  This will be investigated if and when MPAS-Analysis can accommodate a test case with Cartesian coordinates.
+
+<h2>Design solution: Support for arbitrary output interpolation points <br>
+Date last modified: 2017/02/25 <br>
+Contributors: Xylar Asay-Davis
+</h2>
+
+I do not intend to address this consideration in this project.  It may be that `ESMF_RegridWeightGen` can also be used to perform interpolation to arbitrary points (in particular, a set of points that are not cell centers or vertices of a mesh), but this is not yet clear to me.  If not, an alternative solution for arbitrary destination points will be needed.
+
+<h2>Design solution: Support caching results from any costly, one-time geometric computations <br>
+Date last modified: 2017/02/25 <br>
+Contributors: Xylar Asay-Davis</h2>
+
+This should be relatively easy to accommodate with `ESMF_RegridWeightGen` and `ncremap`.  The default behavior of the function for generating interpolation weights will be to do nothing if the mapping file already exists.  Further, we can support an optional config option that will point to an existing mapping file if one has already been generated and cached somewhere (e.g. in a shared directory).  Eventually, we will probably want to systematically store these mapping files for typical MPAS meshes and typical output grids, particularly for those that are expensive to generate.
```

Comment:
@milenaveneziani, that's a little tricky because the same mapping file gets used to interpolate many fields.  We would need to check that they *all* exist before deciding whether to create the mapping file.  This would require looping over the possible output files to check if they all exist, then creating the mapping file if not, then looping over the output files again to produce the interpolation.  This is certainly not impossible to do but it adds a layer of complexity in terms of each script having a list of output files ahead of time, as opposed to generating them on the fly inside a loop.

I also don't see a frequent use case where you would have cached versions of all the interpolated files but somehow would have lost the cached version of the mapping file.  If you *do* think this will be a situation we will frequently find ourselves in, could you please explain a little more how this would happen?